### PR TITLE
internal/graphicsdriver/metal: remove C for macOS

### PR DESCRIPTION
**Pull request by TotallyGamerJet**

### CODE REVIEW

File: internal/cocoasdk/api_coregraphics_darwin.go [hunk 1]
```diff
@@ -0,0 +1,5 @@
+package cocoasdk
```
Comment (hajimehoshi) [inline internal/cocoasdk/api_coregraphics_darwin.go:1]: Add a licence comment.

File: internal/cocoasdk/nsinvocation_darwin.go [hunk 2]
```diff
@@ -0,0 +1,46 @@
+package cocoasdk
```
Comment (hajimehoshi) [inline internal/cocoasdk/nsinvocation_darwin.go:1]: Add a licence comment.

File: internal/cocoasdk/nsinvocation_darwin.go [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+import (
+	"github.com/ebitengine/purego/objc"
+	"unsafe"
```
Comment (hajimehoshi) [inline internal/cocoasdk/nsinvocation_darwin.go:5]:
```
"unsafe"

"github.com/ebitengine/purego/objc"
```

File: internal/graphicsdriver/metal/ca/ca_darwin.go [hunk 4]
```diff
@@ -21,21 +21,16 @@ package ca
...
 import (
 	"errors"
+	"fmt"
+	"runtime"
 	"unsafe"
 
+	"github.com/ebitengine/purego"
+	"github.com/ebitengine/purego/objc"
+	"github.com/hajimehoshi/ebiten/v2/internal/cocoasdk"
 	"github.com/hajimehoshi/ebiten/v2/internal/graphicsdriver/metal/mtl"
```
Comment (hajimehoshi) [inline internal/graphicsdriver/metal/ca/ca_darwin.go:31]: Separate the group for external dependencies and internal dependencies

```
	"github.com/ebitengine/purego"
	"github.com/ebitengine/purego/objc"

	"github.com/hajimehoshi/ebiten/v2/internal/cocoasdk"
	"github.com/hajimehoshi/ebiten/v2/internal/graphicsdriver/metal/mtl"
```

File: internal/cocoasdk/nsmethodsignature_darwin.go [hunk 5]
```diff
@@ -0,0 +1,20 @@
+package cocoasdk
+
+import "github.com/ebitengine/purego/objc"
```
Comment (hajimehoshi) [inline internal/cocoasdk/nsmethodsignature_darwin.go:3]: Use `()`

File: internal/cocoasdk/api_coregraphics_darwin.go [hunk 6]
```diff
@@ -0,0 +1,5 @@
...
+
+type CGSize struct {
+	Width, Height float64
```
Comment (hajimehoshi) [inline internal/cocoasdk/api_coregraphics_darwin.go:4]: Isn't this float32?

File: internal/cocoasdk/nsmethodsignature_darwin.go [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+var class_NSMethodSignature = objc.GetClass("NSMethodSignature\x00")
+
+var (
+	sel_instanceMethodSignatureForSelector = objc.RegisterName("instanceMethodSignatureForSelector:\x00")
+	sel_signatureWithObjCTypes             = objc.RegisterName("signatureWithObjCTypes:\x00")
+)
+
+type NSMethodSignature objc.ID
+
+func InstanceMethodSignatureForSelector(self objc.ID, _cmd objc.SEL) NSMethodSignature {
```
Comment (hajimehoshi) [inline internal/cocoasdk/nsmethodsignature_darwin.go:14]: What about `cmd` instead of `_cmd`?

File: internal/cocoasdk/nsinvocation_darwin.go [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+)
+
+var class_NSInvocation = objc.GetClass("NSInvocation\x00")
```
Comment (hajimehoshi) [inline internal/cocoasdk/nsinvocation_darwin.go:8]: Actually we don't have to use `\x00` but we use `\x00` for performance, right?

If the performance really matters, it's OK, but I feel like this degrates the readability. I prefer not having `\x00`. What do you think?

File: internal/graphicsdriver/metal/ca/ca_darwin.go [hunk 9]
```diff
@@ -106,94 +114,91 @@ func (ml MetalLayer) SetPixelFormat(pf mtl.PixelFormat) {
...
 // are synchronized with the display's refresh rate.
 //
 // Reference: https://developer.apple.com/documentation/quartzcore/cametallayer/2887087-displaysyncenabled.
 func (ml MetalLayer) SetDisplaySyncEnabled(enabled bool) {
-	switch enabled {
-	case true:
-		C.MetalLayer_SetDisplaySyncEnabled(ml.metalLayer, 1)
-	case false:
-		C.MetalLayer_SetDisplaySyncEnabled(ml.metalLayer, 0)
+	if runtime.GOOS == "ios" {
```
Comment (hajimehoshi) [inline internal/graphicsdriver/metal/ca/ca_darwin.go:128]: This would not work with Go 1.15 (`ios` is not one of GOOS yet at that time). Why did you change the logic here by the way?

File: internal/graphicsdriver/metal/ca/ca_darwin.go [hunk 10]
```diff
@@ -106,94 +114,91 @@ func (ml MetalLayer) SetPixelFormat(pf mtl.PixelFormat) {
...
 	}
+	ml.metalLayer.Send(objc.RegisterName("setDisplaySyncEnabled:\x00"), enabled)
 }
 
 // SetDrawableSize sets the size, in pixels, of textures for rendering layer content.
 //
 // Reference: https://developer.apple.com/documentation/quartzcore/cametallayer/1478174-drawablesize.
 func (ml MetalLayer) SetDrawableSize(width, height int) {
-	C.MetalLayer_SetDrawableSize(ml.metalLayer, C.double(width), C.double(height))
+	//TODO: once objc supports calling functions with struct arguments replace this with just a ID.Send call
```
Comment (hajimehoshi) [inline internal/graphicsdriver/metal/ca/ca_darwin.go:138]: add a space after `//`

File: internal/graphicsdriver/metal/view_ios.go [hunk 11]
```diff
@@ -17,6 +17,10 @@
 
 package metal
 
+// Suppress the warnings about availability guard with -Wno-unguarded-availability-new.
+// It is because old Xcode (8 or older?) does not accept @available syntax.
+
+// #cgo CFLAGS: -Wno-unguarded-availability-new
 // #cgo CFLAGS: -x objective-c
```
Comment (hajimehoshi) [inline internal/graphicsdriver/metal/view_ios.go:24]: Unity these two `CFLAGS`.

**Review by hajimehoshi — COMMENTED** [review]
Thank you!

File: internal/cocoasdk/api_coregraphics_darwin.go (hunk 1 above)
Comment (hajimehoshi) [inline internal/cocoasdk/api_coregraphics_darwin.go:1]: I'm not sure, but what about just `cocoa`? Do you have a plan to use other similar packages? I'd like to know your plan.

File: internal/cocoasdk/api_coregraphics_darwin.go (hunk 6 above)
Comment (TotallyGamerJet) [reply internal/cocoasdk/api_coregraphics_darwin.go:4]: CGSize is made up of CGFloat which is a double. [See Apple docs](https://developer.apple.com/documentation/corefoundation/cgfloat?language=objc)

File: internal/cocoasdk/nsmethodsignature_darwin.go (hunk 7 above)
Comment (TotallyGamerJet) [reply internal/cocoasdk/nsmethodsignature_darwin.go:14]: I did _cmd bc that is how it is done in Objc but if you prefer just cmd then I can do that

File: internal/cocoasdk/nsinvocation_darwin.go (hunk 8 above)
Comment (TotallyGamerJet) [reply internal/cocoasdk/nsinvocation_darwin.go:8]: I didn't notice any major differences. Ill remove them

File: internal/cocoasdk/nsmethodsignature_darwin.go (hunk 7 above)
Comment (hajimehoshi) [reply internal/cocoasdk/nsmethodsignature_darwin.go:14]: > I did _cmd bc that is how it is done in Objc

Oh really? Could you show me a link for the explanation?

File: internal/cocoasdk/api_coregraphics_darwin.go (hunk 6 above)
Comment (hajimehoshi) [reply internal/cocoasdk/api_coregraphics_darwin.go:4]: Acknowledged, thanks!

**Comment by hajimehoshi** [comment]
There seems an error in the testing bot:

```
# github.com/hajimehoshi/ebiten/v2/internal/graphicsdriver/metal
Undefined symbols for architecture arm64:
  "_objc_msgSend", referenced from:
      _addSublayer in _x002.o
      _setFrame in _x002.o
ld: symbol(s) not found for architecture arm64
```

File: internal/cocoasdk/api_coregraphics_darwin.go (hunk 1 above)
Comment (TotallyGamerJet) [reply internal/cocoasdk/api_coregraphics_darwin.go:1]: Cocoa is really just a collection of most of the graphical frameworks like AppKit, Foundation etc. I think just cocoa is fine as the name. The goal is to port enough to make all of ebitengine including glfw into purego.

File: internal/graphicsdriver/metal/ca/ca_darwin.go (hunk 9 above)
Comment (TotallyGamerJet) [reply internal/graphicsdriver/metal/ca/ca_darwin.go:128]: The logic should be the same. There was a check in the C code to only run it on IOS. I replaced all runtime.GOOS == "iOS" with just a check if the Object responds to a certain function.

File: internal/cocoasdk/nsmethodsignature_darwin.go (hunk 7 above)
Comment (TotallyGamerJet) [reply internal/cocoasdk/nsmethodsignature_darwin.go]: The reason is likely because they didn't want to take up the identifier `cmd` from use in a method. This [Apple Doc](https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/ObjCRuntimeGuide/Articles/ocrtHowMessagingWorks.html) says "The first two arguments passed to the procedure are the receiving object (self) and the method selector (_cmd). These arguments are hidden in method syntax but must be made explicit when the method is called as a function." Although it doesn't say _why_ it is this way.

Comment (hajimehoshi) [reply internal/cocoasdk/nsmethodsignature_darwin.go]: OK I'm convinced.

File: internal/cocoasdk/api_coregraphics_darwin.go (hunk 1 above)
Comment (hajimehoshi) [reply internal/cocoasdk/api_coregraphics_darwin.go]: I mean I wanted to know you would add other packages besides `cocoa` and I wanted to know the naming convention. Anway we can revisit this later. Let's go with `cocoa`.

File: internal/graphicsdriver/metal/ca/ca_darwin.go (hunk 9 above)
Comment (hajimehoshi) [reply internal/graphicsdriver/metal/ca/ca_darwin.go]: The latest change (checking the existence of the method) seems fine.

**Review by hajimehoshi — COMMENTED** [review]
Let me test your PR on my MacBook Pro and iOS. Thanks!

**Comment by hajimehoshi** [comment]
I found https://github.com/hajimehoshi/ebiten/pull/2243#issuecomment-1213639418 still exists. Please take a look, thanks

**Comment by TotallyGamerJet** [comment]
yeah it wasn't linking to QuartzCore since that got removed from metal and wasn't in the iOS specific code.

Should hopefully be fixed now.

**Comment by hajimehoshi** [comment]
Other errors:

```
cannot use float64(width) (type float64) as type cocoa.CGFloat in field value
```

**Comment by TotallyGamerJet** [comment]
Hmm I’m not sure why it’s complaining on iOS although I did just remember I was using a modified version of purego so I’ll create that PR tomorrow morning

**Comment by hajimehoshi** [comment]
> Hmm I’m not sure why it’s complaining on iOS although I did just remember I was using a modified version of purego so I’ll create that PR tomorrow morning

So, we have to update purego in go.mod in this PR, right?

**Comment by TotallyGamerJet** [comment]
Yes

**Comment by hajimehoshi** [comment]
OK so are you updating go.mod in this PR? The test bots seem happy already.

**Comment by TotallyGamerJet** [comment]
It should all be working now!

**Comment by hajimehoshi** [comment]
Let me test with actual devices tomorrow. Thank you for your hard work!

File: internal/cocoa/nsmethodsignature_darwin.go [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+	sel_signatureWithObjCTypes             = objc.RegisterName("signatureWithObjCTypes:")
+)
+
+type NSMethodSignature objc.ID
+
+func InstanceMethodSignatureForSelector(self objc.ID, _cmd objc.SEL) NSMethodSignature {
+	return NSMethodSignature(self.Send(sel_instanceMethodSignatureForSelector, _cmd))
+}
+
+func NSMethodSignature_SignatureWithObjCTypes(types string) NSMethodSignature {
```
Comment (hajimehoshi) [inline internal/cocoa/nsmethodsignature_darwin.go:34]: I think both `InstanceMethodSignatureForSelector` and `SignatureWithObjCTypes` are static methods of NSMethodSignature. So, shouldn't they have consist names like `NSMethodSignature_InstanceMethodSignatureForSelector` and `NSMethodSignature_SignatureWithObjCTypes` (as it is)?

File: internal/graphicsdriver/metal/graphics_darwin.go [hunk 13]
```diff
@@ -21,27 +21,27 @@ import (
...
-// }
-//
-// static void releaseAutoreleasePool(void* pool) {
-//   [(NSAutoreleasePool*)pool release];
-// }
-import "C"
+var (
+	class_NSAutoreleasePool = objc.GetClass("NSAutoreleasePool\x00")
+	sel_new                 = objc.RegisterName("new\x00")
+	sel_release             = objc.RegisterName("release\x00")
```
Comment (hajimehoshi) [inline internal/graphicsdriver/metal/graphics_darwin.go:35]: Remove `\x00`

**Comment by hajimehoshi** [comment]
[go-inovation](https://github.com/hajimehoshi/go-inovation) crashed with this log:

```
2022-08-14 13:27:24.180272+0900 goinovation[407:9762] Metal API Validation Enabled
2022-08-14 13:27:24.363173+0900 goinovation[407:9762] -[AGXA13Device isHeadless]: unrecognized selector sent to instance 0x14a008800
2022-08-14 13:27:24.363780+0900 goinovation[407:9762] *** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: '-[AGXA13Device isHeadless]: unrecognized selector sent to instance 0x14a008800'
*** First throw call stack:
(0x1e0cd8288 0x1f9a07744 0x1e0db5fc0 0x1e0c6cba0 0x1e0c6bf70 0x10468cee4 0x1045f188c)
libc++abi: terminating with uncaught exception of type NSException
dyld4 config: DYLD_LIBRARY_PATH=/usr/lib/system/introspection DYLD_INSERT_LIBRARIES=/Developer/usr/lib/libBacktraceRecording.dylib:/Developer/usr/lib/libMainThreadChecker.dylib:/Developer/Library/PrivateFrameworks/DTDDISupport.framework/libViewDebuggerSupport.dylib:/Developer/Library/PrivateFrameworks/GPUTools.framework/libglInterpose.dylib
*** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: '-[AGXA13Device isHeadless]: unrecognized selector sent to instance 0x14a008800'
terminating with uncaught exception of type NSException
(
```

**Comment by hajimehoshi** [comment]
On an iOS simulator, go-inovation crashed at 

```go
		colorspace, _, _ := purego.SyscallN(_CGColorSpaceCreateWithName, **(**uintptr)(unsafe.Pointer(&kCGColorSpaceDisplayP3))) // Dlsym returns pointer to symbol so dereference it
```

with EXC_BAD_ACCESS.

**Comment by TotallyGamerJet** [comment]
Both of the crashes you experienced are from calling functions that in the .m files were marked as not to run on iOS Those are the ones that I had replaced with `if runtime.GOOS != "ios"`. I'm not sure of any other ways to not call them on iOS since they appear to be lying that they can be called. How do you want to proceed?

**Comment by hajimehoshi** [comment]
If we just want to distinguish ios or not, what about making these files?

```go
//go:build ios
// +build ios

const isIos = true
```

```go
//go:build !ios
// +build !ios

const isIos = false
```

**Comment by hajimehoshi** [comment]
Or we can move the method to these files.

**Comment by TotallyGamerJet** [comment]
The build tag method I think is best

**Comment by hajimehoshi** [comment]
Both (defining consts and defining functions) use build tags. You mean you prefer defining functions with build tags? That's fine :-)

**Comment by TotallyGamerJet** [comment]
Uh no sorry. I meant I like the consts one. What I wrote was very unclear

**Comment by hajimehoshi** [comment]
> Uh no sorry. I meant I like the consts one. What I wrote was very unclear

No problem, the const way is also fine!

File: internal/graphicsdriver/metal/ca/ca_darwin.go [hunk 14]
```diff
@@ -49,40 +45,51 @@ type Layer interface {
...
+	kCGColorSpaceDisplayP3      = purego.Dlsym(coreGraphics, "kCGColorSpaceDisplayP3")
+)
+
 // MakeMetalLayer creates a new Core Animation Metal layer.
 //
 // Reference: https://developer.apple.com/documentation/quartzcore/cametallayer.
 func MakeMetalLayer() MetalLayer {
-	return MetalLayer{C.MakeMetalLayer()}
+	layer := objc.ID(objc.GetClass("CAMetalLayer")).Send(objc.RegisterName("new"))
+	if !cocoa.IsIOS {
```
Comment (hajimehoshi) [inline internal/graphicsdriver/metal/ca/ca_darwin.go:63]: I think this constant should be defined in ca as long as this is the only one usage, but what do you think?

File: internal/cocoa/isIOS.go [hunk 15]
```diff
@@ -1,4 +1,4 @@
-// Copyright 2018 The Ebiten Authors
+// Copyright 2022 The Ebitengine Authors
```
Comment (hajimehoshi) [inline internal/cocoa/isIOS.go:1]: Use a small cases for a file name.

Comment (hajimehoshi) [reply internal/cocoa/isIOS.go:1]: I think `ios.go` is fine.

File: internal/cocoa/notIOS.go [hunk 16]
```diff
@@ -0,0 +1,20 @@
+// Copyright 2022 The Ebitengine Authors
```
Comment (hajimehoshi) [inline internal/cocoa/notIOS.go:1]: ditto: please use small cases. `notios.go` is fine. 

Unfortunately it might be hard to change only cases of a file name. You might have to remove this file once, commit the change, then add the file again.

File: internal/graphicsdriver/metal/ca/ca_darwin.go (hunk 14 above)
Comment (TotallyGamerJet) [reply internal/graphicsdriver/metal/ca/ca_darwin.go:63]: It's also used in mtl package

Comment (hajimehoshi) [reply internal/graphicsdriver/metal/ca/ca_darwin.go:63]: NVM, I found it https://github.com/hajimehoshi/ebiten/pull/2243/files#diff-f436c80b858b5bc84325c6aa909731c966b62551cefefc37b6f6936ab48dccd5R584

**Comment by hajimehoshi** [comment]
I'll test this with actual devices tomorrow, thanks!

**Comment by hajimehoshi** [comment]
It seems like an iOS emulator and an actual iOS device worked with this PR. I'll double-check this PR as this is pretty big.

File: internal/cocoa/ios.go [hunk 17]
```diff
@@ -1,4 +1,4 @@
-// Copyright 2018 The Ebiten Authors
+// Copyright 2022 The Ebitengine Authors
```
Comment (hajimehoshi) [inline internal/cocoa/ios.go:1]: Sorry but I realized that `ios_darwin.go` is better so that non-Darwn environment cannot build this package.

File: internal/cocoa/notios.go [hunk 18]
```diff
@@ -0,0 +1,20 @@
+// Copyright 2022 The Ebitengine Authors
```
Comment (hajimehoshi) [inline internal/cocoa/notios.go:1]: ditto: `notios_darwin.go` is better.

File: internal/graphicsdriver/metal/ca/ca_darwin.go [hunk 19]
```diff
@@ -106,94 +115,91 @@ func (ml MetalLayer) SetPixelFormat(pf mtl.PixelFormat) {
...
 // are synchronized with the display's refresh rate.
 //
 // Reference: https://developer.apple.com/documentation/quartzcore/cametallayer/2887087-displaysyncenabled.
 func (ml MetalLayer) SetDisplaySyncEnabled(enabled bool) {
-	switch enabled {
-	case true:
-		C.MetalLayer_SetDisplaySyncEnabled(ml.metalLayer, 1)
-	case false:
-		C.MetalLayer_SetDisplaySyncEnabled(ml.metalLayer, 0)
+	if runtime.GOOS == "ios" {
```
Comment (hajimehoshi) [inline internal/graphicsdriver/metal/ca/ca_darwin.go:129]: `cocoa.IsIOS`?

File: internal/graphicsdriver/metal/graphics_darwin.go [hunk 20]
```diff
@@ -21,27 +21,27 @@ import (
 	"strings"
 	"unsafe"
 
+	"github.com/ebitengine/purego/objc"
 	"github.com/hajimehoshi/ebiten/v2/internal/graphics"
```
Comment (hajimehoshi) [inline internal/graphicsdriver/metal/graphics_darwin.go:25]: Separater external and internal dependencies i.e. add an empty line after `"github.com/ebitengine/purego/objc"`

File: internal/graphicsdriver/metal/mtl/mtl_darwin.go [hunk 21]
```diff
@@ -338,6 +337,14 @@ const (
 	CommandBufferStatusError       CommandBufferStatus = 5 // Execution of the command buffer was aborted due to an error during execution.
 )
 
+var (
+	metal                         = purego.Dlopen("Metal.framework/Metal", purego.RTLD_GLOBAL)
+	_MTLCreateSystemDefaultDevice = purego.Dlsym(metal, "MTLCreateSystemDefaultDevice")
+
+	_ = purego.Dlopen("/System/Library/Frameworks/CoreGraphics.framework/Versions/Current/CoreGraphics", purego.RTLD_GLOBAL)
```
Comment (hajimehoshi) [inline internal/graphicsdriver/metal/mtl/mtl_darwin.go:344]: Why do we need `Dlopen` without using the returning value here? If we expect only side effects, what about moving this into `init` functions?

File: internal/graphicsdriver/metal/mtl/mtl_darwin.go [hunk 22]
```diff
@@ -457,134 +464,257 @@ type Device struct {
...
+
+func _NSStringtoGoString(nsstring objc.ID) string {
+	// this will be nicer with unsafe.Slice once ebitengine requires 1.17
+	// reflect.SliceHeader is used because it will force Go to copy the string
+	// into Go memory when casted to a string
+	var s []byte
+	header := (*reflect.SliceHeader)(unsafe.Pointer(&s))
+	header.Data = uintptr(nsstring.Send(sel_UTF8String))
+	header.Len = int(nsstring.Send(sel_length))
+	return string(s)
```
Comment (hajimehoshi) [inline internal/graphicsdriver/metal/mtl/mtl_darwin.go:479]: Set `Cap` too.

File: internal/graphicsdriver/metal/mtl/mtl_darwin.go [hunk 23]
```diff
@@ -814,33 +965,57 @@ type BlitCommandEncoder struct {
 //
 // Reference: https://developer.apple.com/documentation/metal/mtlblitcommandencoder/1400775-synchronize.
 func (bce BlitCommandEncoder) Synchronize(resource Resource) {
-	C.BlitCommandEncoder_Synchronize(bce.commandEncoder, resource.resource())
+	if runtime.GOOS == "ios" {
```
Comment (hajimehoshi) [inline internal/graphicsdriver/metal/mtl/mtl_darwin.go:968]: `cocoa.IsIOS`

File: internal/graphicsdriver/metal/mtl/mtl_darwin.go [hunk 24]
```diff
@@ -814,33 +965,57 @@ type BlitCommandEncoder struct {
...
+		return
+	}
+	objc.ID(bce.commandEncoder).Send(sel_synchronizeResource, resource.resource())
 }
 
 func (bce BlitCommandEncoder) SynchronizeTexture(texture Texture, slice int, level int) {
-	C.BlitCommandEncoder_SynchronizeTexture(bce.commandEncoder, texture.texture, C.uint_t(slice), C.uint_t(level))
+	if runtime.GOOS == "ios" {
```
Comment (hajimehoshi) [inline internal/graphicsdriver/metal/mtl/mtl_darwin.go:975]: `cocoa.IsIOS`

File: internal/graphicsdriver/metal/mtl/mtl_darwin.go [hunk 25]
```diff
@@ -849,103 +1024,130 @@ func (l Library) MakeFunction(name string) (Function, error) {
...
 }
 
 // GetBytes copies a block of pixels from the storage allocation of texture
 // slice zero into system memory at a specified address.
 //
 // Reference: https://developer.apple.com/documentation/metal/mtltexture/1515751-getbytes.
 func (t Texture) GetBytes(pixelBytes *byte, bytesPerRow uintptr, region Region, level int) {
-	r := region.c()
-	C.Texture_GetBytes(t.texture, unsafe.Pointer(pixelBytes), C.size_t(bytesPerRow), r, C.uint_t(level))
+	inv := cocoa.NSInvocation_invocationWithMethodSignature(cocoa.NSMethodSignature_SignatureWithObjCTypes("v@:^vQ{MTLRegion={MTLOrigin=qqq}{MTLSize=qqq}}Q"))
```
Comment (hajimehoshi) [inline internal/graphicsdriver/metal/mtl/mtl_darwin.go:1047]: Just out of curiosity, is there a document about this signature string rule?

Comment (TotallyGamerJet) [reply internal/graphicsdriver/metal/mtl/mtl_darwin.go:1047]: Yeah it’s what’s returned from [@encode](https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/ObjCRuntimeGuide/Articles/ocrtTypeEncodings.html#//apple_ref/doc/uid/TP40008048-CH100). Obviously the best thing is to get the signature from the method itself but that doesn’t work with protocols bc the only way is with the protocol_getDescription function which returns a struct. We don’t have struct return support yet which is the reason I’m using NSInvocation.

Comment (hajimehoshi) [reply internal/graphicsdriver/metal/mtl/mtl_darwin.go:1047]: Thanks!

Comment (hajimehoshi) [reply internal/graphicsdriver/metal/mtl/mtl_darwin.go:1047]: > We don’t have struct return support yet which is the reason I’m using NSInvocation.

I think this is a good point to leave a comment.

Comment (TotallyGamerJet) [reply internal/graphicsdriver/metal/mtl/mtl_darwin.go:1047]: Sure I leave it by use of nsinvocation or just at its definition?

Comment (hajimehoshi) [reply internal/graphicsdriver/metal/mtl/mtl_darwin.go:1047]: just at its definition seems enough, thanks!

File: internal/graphicsdriver/metal/view_ios.go [hunk 26]
```diff
@@ -17,8 +17,11 @@
 
 package metal
 
-// #cgo CFLAGS: -x objective-c
-// #cgo LDFLAGS: -framework UIKit
+// Suppress the warnings about availability guard with -Wno-unguarded-availability-new.
```
[the remark this answers is not in the document]
Comment (hajimehoshi) [reply internal/graphicsdriver/metal/view_ios.go:20]: I'm fine with this change in another later PR. How big will the change be?

[the remark this answers is not in the document]
Comment (TotallyGamerJet) [reply internal/graphicsdriver/metal/view_ios.go:20]: Not big but let’s just save it for another PR. This one is big enough

[the remark this answers is not in the document]
Comment (hajimehoshi) [reply internal/graphicsdriver/metal/view_ios.go:20]: I agree. I'm looking forward to your next PR :-)

File: go.mod [hunk 27]
```diff
@@ -3,7 +3,7 @@ module github.com/hajimehoshi/ebiten/v2
 go 1.15
 
 require (
-	github.com/ebitengine/purego v0.0.0-20220729024107-78cdc2949de6
+	github.com/ebitengine/purego v0.0.0-20220813165658-375146121d78
```
Comment (hajimehoshi) [inline go.mod:6]: Update to 8dbe340b03f11115e843b9ee32252727985fb822

**Comment by TotallyGamerJet** [comment]
I would test on actual device again since I did a pretty big refactor. I moved the common OBJC objects into the cocoa package because they will be needed when ports the rest of Ebitengine

File: internal/graphicsdriver/metal/mtl/example_darwin_test.go [hunk 28]
```diff
@@ -17,6 +17,7 @@ package mtl_test
 import (
 	"encoding/json"
 	"fmt"
+	"github.com/ebitengine/purego"
```
Comment (hajimehoshi) [inline internal/graphicsdriver/metal/mtl/example_darwin_test.go:20]: Move this to the same group as `"golang.org/x/image/math/f32"`

**Review by hajimehoshi — APPROVED** [review]
LGTM, thanks!

File: internal/graphicsdriver/metal/mtl/example_darwin_test.go (hunk 28 above)
Comment (superloach) [reply internal/graphicsdriver/metal/mtl/example_darwin_test.go]: Would using something like `gofumpt` help here? IIRC it automatically groups imports like this. Not sure if it matches existing Ebitengine code style though.

File: internal/graphicsdriver/metal/mtl/mtl_darwin.go [hunk 29]
```diff
@@ -457,134 +461,236 @@ type Device struct {
...
-		FrontFaceStencilDepthFailureOperation:     C.uint8_t(dsd.FrontFaceStencil.DepthFailureOperation),
-		FrontFaceStencilDepthStencilPassOperation: C.uint8_t(dsd.FrontFaceStencil.DepthStencilPassOperation),
-		FrontFaceStencilStencilCompareFunction:    C.uint8_t(dsd.FrontFaceStencil.StencilCompareFunction),
-	}
+	depthStencilDescriptor := objc.ID(class_MTLDepthStencilDescriptor).Send(sel_new)
+	backFaceStencil := depthStencilDescriptor.Send(sel_backFaceStencil)
+	backFaceStencil.Send(sel_setStencilFailureOperation, uintptr(dsd.BackFaceStencil.StencilFailureOperation))
+	backFaceStencil.Send(sel_setDepthFailureOperation, uintptr(dsd.BackFaceStencil.DepthFailureOperation))
+	backFaceStencil.Send(sel_setDepthStencilPassOperation, uintptr(dsd.BackFaceStencil.DepthStencilPassOperation))
+	backFaceStencil.Send(sel_setStencilCompareFunction, uintptr(dsd.BackFaceStencil.StencilCompareFunction)) //TODO
```
Comment (hajimehoshi) [inline internal/graphicsdriver/metal/mtl/mtl_darwin.go:684]: What is this TODO? @TotallyGamerJet

Comment (TotallyGamerJet) [reply internal/graphicsdriver/metal/mtl/mtl_darwin.go:684]: It can be removed. It was needed when I was working on it and just never got deleted.

Comment (hajimehoshi) [reply internal/graphicsdriver/metal/mtl/mtl_darwin.go:684]: Sure!

File: internal/graphicsdriver/metal/mtl/example_darwin_test.go (hunk 28 above)
Comment (hajimehoshi) [reply internal/graphicsdriver/metal/mtl/example_darwin_test.go]: I'll take a look later, thanks
